sound/soc/sof/topology.c

```diff
@@ -1233,6 +1240,60 @@ static int sof_control_unload(struct snd_soc_component *scomp,
  * DAI Topology
  */
 
+static int sof_core_enable(struct snd_sof_dev *sdev, int core)
+{
+	struct sof_ipc_pm_core_config pm_core_config = {
+		.hdr = {
+			.cmd = SOF_IPC_GLB_PM_MSG | SOF_IPC_PM_CORE_ENABLE,
+			.size = sizeof(pm_core_config),
+		},
+		.enable_mask = sdev->enabled_cores_mask | 1 << core,
+	};
+	int ret;
+
+	if (sdev->enabled_cores_mask & 1 << core)
+		return 0;
+
+	/* power up the core */
+	ret = snd_sof_dsp_core_power_up(sdev, 1 << core);
+	if (ret < 0) {
+		dev_err(sdev->dev, "error: %d powering up core %d\n",
+			ret, core);
+		return ret;
+	}
+
+	/* update enabled cores mask */
+	sdev->enabled_cores_mask |= 1 << core;
+
+	/* Now notify DSP that the core has been powered up */
+	ret = sof_ipc_tx_message(sdev->ipc, pm_core_config.hdr.cmd,
+				 &pm_core_config, sizeof(pm_core_config),
+				 &pm_core_config, sizeof(pm_core_config));
+	if (ret < 0)
+		dev_err(sdev->dev, "error: core %d enable ipc failure %d\n",
+			core, ret);
+
+	return ret;
+}
+
+int sof_pipeline_core_enable(struct snd_sof_dev *sdev,
+			    const struct snd_sof_widget *swidget)
+{
+	const struct sof_ipc_pipe_new *pipeline;
+	int core;
+
+	if (swidget->id == snd_soc_dapm_scheduler) {
+		pipeline = swidget->private;
+	} else {
+		pipeline = sof_pipeline_find(sdev, swidget->pipeline_id);
+		if (!pipeline)
+			return -ENOENT;
+	}
+	core = pipeline->core;
+
+	return sof_core_enable(sdev, core);
```

Comment:
@lyakh i could be wrong but wouldnt this be a duplicate call to sof_core_enable(). We already have a sof_core_enable() in sof_widget_ready()?

Reply:
@ranj063 well, I don't know. What if in a pipeline all components are placed on core 0 and the scheduler widget has 0 for component core and 1 for scheduling core? This is certainly a weird example. Besides, that function is extremely light-weight if the core is already enabled.

Reply:
@lyakh @ranj063 If there's a theoretical chance, let's keep it. It's not like we have rules laid out anywhere currently, so this could come and haunt us later.

Reply:
@lyakh @kv2019i  I would argue the other way around. In the past, we have seen FE hangs because we tried to power up a core that was already up. And because you call snd_sof_dsp_power_up() without checking if it is already up, I am afraid we could run into similar issues again. So even if there is a slight chance that this could happen, we should prevent it

Reply:
@ranj063 there is a *VERY* explicit check for that:
```
	if (sdev->enabled_cores_mask & BIT(core))
		return 0;
```
or do you see any situations where this might not trigger?

Reply:
@lyakh sorry I completely missed this check. I only looked in the core_power_up() op to see if the check was there :)

Reply:
> @ranj063 well, I don't know. What if in a pipeline all components are placed on core 0 and the scheduler widget has 0 for component core and 1 for scheduling core? This is certainly a weird example. Besides, that function is extremely light-weight if the core is already enabled.

@lyakh this would be very weird. Besides for my education, what is the real difference between a comp core and a schedule core? Is a schedule core the core that the pipeline is supposed to run on? Is it possible for the comp core for the components in the pipeline to be different from the pipeline sched core at all?

Reply:
@ranj063 As far as I understand originally only a pipeline-level multi-core support was implemented, i.e. you could only specify cores per pipeline. That's what the scheduling core is for. Recently a capability has been added to run individual components on different cores. I still don't know what use-cases this should be helpful for. I asked for examples, but I don't think I've got any back. So, now in theory it should be possible to run a pipeline on slave core 1 and a certain component from it on core 2. That means we now have to specify cores per component for each component.
As for this specific examples - all components are marked for cores other than the scheduling core - as I said, I don't know if this is meaningful at all. In theory you can of course imagine such cases: you can have your timer interrupt for the pipeline fire on one core and all DMA copying done on other cores. On our current hardware I'm not aware of any good reasons to do that.